fpga_interchange/yosys_json.py

```diff
@@ -0,0 +1,417 @@
+
+    >>> bits = [2, 3, 4, 5]
+    >>> tuple(interp_yosys_net(bits, offset=0, upto=0))
+    ((0, 2), (1, 3), (2, 4), (3, 5))
+    >>> tuple(interp_yosys_net(bits, offset=4, upto=0))
+    ((4, 2), (5, 3), (6, 4), (7, 5))
+    >>> tuple(interp_yosys_net(bits, offset=0, upto=1))
+    ((3, 2), (2, 3), (1, 4), (0, 5))
+    >>> tuple(interp_yosys_net(bits, offset=4, upto=1))
+    ((7, 2), (6, 3), (5, 4), (4, 5))
+
+    """
+    if not upto:
+        for idx, bit in enumerate(bits):
+            yield idx + offset, bit
+    else:
+        for idx, bit in enumerate(bits):
+            yield offset + (len(bits) - 1) - idx, bit
+
+
+# FIXME: These constants should likely be from the device database.
+VCC_CELL_TYPE = 'VCC'
+GND_CELL_TYPE = 'GND'
+VCC_PORT = 'P'
+GND_PORT = 'G'
+
+
+def create_unique_name(names, base_name):
+    if base_name not in names:
+        return base_name
+
+    idx = 0
+    while True:
+        name = '{}_{}'.format(base_name, idx)
+        if name not in names:
+            return name
+
+        idx += 1
+
+
+def make_vcc_cell(cell, module_data):
+    cell_names = set(cell.cell_instances.keys()) | set(
+        module_data['cells'].keys())
+    vcc_cell = create_unique_name(cell_names, '$__vcc')
+    cell.add_cell_instance(name=vcc_cell, cell_name=VCC_CELL_TYPE)
+
+    return vcc_cell
+
+
+def make_gnd_cell(cell, module_data):
+    cell_names = set(cell.cell_instances.keys()) | set(
+        module_data['cells'].keys())
+    gnd_cell = create_unique_name(cell_names, '$__gnd')
+    cell.add_cell_instance(name=gnd_cell, cell_name=GND_CELL_TYPE)
+
+    return gnd_cell
+
+
+def convert_cell(module_name, module_data, library, libraries, modules,
+                 verbose, errors):
+    for cell_name, cell_data in module_data['cells'].items():
+        # Don't import modules that are missing children, they likely aren't
+        # important.
+        if cell_data['type'] not in modules:
+            errors[
+                module_name] = 'Failed to import cell type {} because it has a undefined cell {}'.format(
+                    module_name, cell_data['type'])
+            return
+
+    property_map = {}
+    if 'attributes' in module_data:
+        property_map = module_data['attributes']
+
+    cell = Cell(module_name, property_map)
+    libraries[library].add_cell(cell)
+
+    net_duplicate_names = {}
+
+    def add_net(net_name, net_data, track_duplicates=False):
+        property_map = {}
+        if 'attributes' in net_data:
+            property_map.update(net_data['attributes'])
+
+        offset = net_data.get('offset', 0)
+        upto = net_data.get('upto', 0)
+        if is_bus(net_data['bits'], offset, upto):
+            for bit_index, bit in interp_yosys_net(net_data['bits'], offset,
+                                                   upto):
+                name = '{}[{}]'.format(net_name, bit_index)
+                cell.add_net(name, property_map)
+
+                if bit == '0':
```

Comment:
'x' and 'z' bits can also occur, in this context I think converting these to disconnected ports for both should be OK

Reply:
`'x'` and `'z'` bits need to be handled prior to the FPGA interchange logical netlist format.  For example my murax yosys script is currently:

```
yosys -import

read_verilog Murax.v
read_verilog basys3_toplevel.v

synth_xilinx -nolutram -nowidelut -nosrl -nocarry -nodsp

# opt_expr -undriven makes sure all nets are driven, if only by the $undef
# net.
opt_expr -undriven
opt_clean

setundef -zero -params

write_json murax.json
```

So while `'x'` and `'z'` bits legal in Yosys Netlist JSON, they are not legal as input to the FPGA interchange.